### ISSUE TITLE
replace keypad workarounds

### ISSUE THREAD

**Issue by pixelzoom** [phetsims/projectile-motion#127]
Motivated by https://github.com/phetsims/projectile-motion/issues/118.

KeypadLayer and KeypadPanel are a workaround copied from unit-rates.  They are a workaround for the lack of a robust Dialog (phetsims/joist#359) and the lack of closure on the next-gen keypad (phetsims/scenery-phet#283, phetsims/scenery-phet#271). 

Rather than factor out "workaround" code, KeypadLayer and KeypadPanel should be replaced when the above-mentioned issues are addressed.  

Given the rate of progress on those issues, they are unlikely to be addressed any time soon.  So this issue is not a requirement for 1.0.0 publication.  And this issues is being marked as deferred.

**Comment by pixelzoom** [comment]
7/27/17 dev meeting:  @jonathanolson is using the next-gen keypad in area-model, so @andrea-phet should be able to use it in projectile-motion too.  So replace KeypadLayer and KeypadPanel with the next-gen stuff prior to the 1.0.0 release.  

@andrea-phet please work with @jbphet and @jonathanolson on this.

**Comment by pixelzoom** [comment]
@andrea-phet @jbphet @jnnathanolson whatever you do in projectile-motion should be usable in unit-rates, see https://github.com/phetsims/unit-rates/issues/153.

**Comment by jbphet** [comment]
@andrea-phet - I've updated and cleaned up the common keypad code, so please go ahead and use it in Projectile Motion.  The code is in scenery-phet/js/keypad and there are several demos in scenery-phet/js/demo.  Also, it is currently being used in Area Model Multiplication simulation, so you can find additional examples there if needed.

**Comment by andrea-phet** [comment]
Assigning to @jbphet to review.

**Comment by pixelzoom** [comment]
I took a quick peek at https://github.com/phetsims/projectile-motion/commit/c4957e3b9cd59642e64d1123f3c0b136258cb6b8.  This address one part of the duplicated code, by replacing KeypadPanel with scenery-phet.Keypad.   We'll still need to decide what to do about KeypadLayer, which was also copied from unit-rates.

**Comment by pixelzoom** [comment]
KeypadLayer still refers to KeypadPanel:

```js
44 // KeypadPanel options
```

**Comment by pixelzoom** [comment]
This issue should not be closed until Keypad.dispose is implemented, see https://github.com/phetsims/scenery-phet/issues/283#issuecomment-320370502.

**Comment by pixelzoom** [comment]
I fixed the comments mentioned in https://github.com/phetsims/projectile-motion/issues/127#issuecomment-321379590.

**Comment by pixelzoom** [comment]
The integration of scenery.Keypad looks good, so that bit is done.  But as I mentioned in https://github.com/phetsims/projectile-motion/issues/127#issuecomment-321378867, we need to decide what to do about the duplication in KeypadLayer (which was also copied from unit-rates).

KeypadLayer (in unit-rates) was a workaround for deficiencies and bugs in joist.Dialog. I'm not confident that Dialog is up to the task yet, since it's still being redesigned.  And we would still need the parts of KeypadLayer that handle positioning the Dialog.  It also looks like KeypadLayer has been modified somewhat from unit-rates, and it's difficult for me to evaluate how much code is still shared until https://github.com/phetsims/unit-rates/issues/153 is completed.

So... My recommendation is:

(1) Fix the remaining issues in Keypad, as noted in phetsims/scenery-phet#283.  This is a prerequisite to publishing 1.0.0. 

(2) Leave this issue open, and use Dialog when it has been revamped.  This is not required for publishing 1.0.0.

**Comment by pixelzoom** [comment]
@jbphet I'm labeling this issue as "blocked" until you complete work on Keypad as described in  phetsims/scenery-phet#283.

**Comment by jbphet** [comment]
The work on keypad is essentially complete.  There is a test case for dispose of the keypad that is failing, and I'll follow up on that, but this shouldn't block this sim since `Keypad.dispose` is not used.

**Comment by andrea-phet** [comment]
Leaving this issue open for when Dialog is revamped, but this does not need to be addressed before publishing 1.0.0.

**Comment by pixelzoom** [comment]
@jbphet said:
> The work on keypad is essentially complete. There is a test case for dispose of the keypad that is failing, and I'll follow up on that, but this shouldn't block this sim since Keypad.dispose is not used.

Depending on the resolution to https://github.com/phetsims/projectile-motion/issues/135, Keypad.dispose may in fact be required by this sim.   And it's certainly required to proceed with Unit Rates in https://github.com/phetsims/unit-rates/issues/153.

**Comment by pixelzoom** [comment]
To clarify my comment in https://github.com/phetsims/projectile-motion/issues/127#issuecomment-321933453 ...  This sim currently uses the same Keypad configuration for entry of all values.  That means that the user can enter some values that aren't valid for a field.  In Unit Rates, a new keypad is used for each field, with options tailored to that field.

**Comment by jbphet** [comment]
The failing test case for the Keypad dispose function is now resolved, see https://github.com/phetsims/scenery-phet/issues/283 and https://github.com/phetsims/scenery-phet/issues/330.  It should be safe to use a version of the keypad that is added and removed if the design moves that direction.  Unassigning myself, @andrea-phet - feel free to assign back if there is any further work needed here.

**Comment by andrea-phet** [comment]
Okay. Again, leaving this issue open but unassigned in case Dialog should be used later.

**Comment by pixelzoom** [comment]
Changing this from on-hold to deferred, since no further work is required for 1.0.0 release.

**Comment by zepumph** [comment]
Pinging @Denz1994 for current progress on KeyPad. Is this something that is close enough to completion that we should add it before instrumenting the workarounds for PhET-iO. That would definitely be preferred.

**Comment by samreid** [comment]
@zepumph can you please re-evaluate and determine assignee?

**Comment by zepumph** [comment]
I feel like we could either wait on https://github.com/phetsims/scenery-phet/issues/283, or decide that keypad is good enough to add into this release.

**Comment by zepumph** [comment]
This may be getting some traction because https://github.com/phetsims/scenery-phet/issues/283 is a quarterly goal right now. Assigning myself.

**Comment by zepumph** [comment]
Having looked at this issue in its entirety, the majority of work seems to be in the `KeypadLayer` that is duplicated here and in unit rates, and less about the use of Keypad.

**Comment by jbphet** [comment]
@zepumph is going to take this on (thanks @zepumph!), I'm unassigning myself for now.

**Comment by matthew-blackman** [comment]
Any updates on this? Seems like it is related to #307.

**Comment by zepumph** [comment]
I think it will. We still haven't had a push to centralize a keypad container (especially a modal one) for usages like this sim and unit rates. When we do the hacky, sim-specific KeypadLayer should be replaced with common code.

**Comment by zepumph** [comment]
We now have SCENERY_PHET/KeypadDialog. I think that is a good candidate for initial exploration in this issue.
